src/rgw/rgw_sts.cc

```diff
@@ -0,0 +1,129 @@
+  calc_hmac_sha256(key.c_str(), key.size(), string_to_sign.c_str(), string_to_sign.size(), signature);
+
+  char aux[CEPH_CRYPTO_HMACSHA256_DIGESTSIZE * 2 + 1];
+  buf_to_hex((unsigned char *) signature, CEPH_CRYPTO_HMACSHA256_DIGESTSIZE, aux);
+  string signature_str = string(aux);
+  session_token = rgw::to_base64(signature_str);
+}
+
+void help()
+{
+  cout<<"option:  "<<"--help  help"<< std::endl;
+  cout<<"         "<<"--role-id  role id"<< std::endl;
+  cout<<"         "<<"--policy  policy"<< std::endl;
+  cout<<"         "<<"--saml-keys  saml keys"<< std::endl;
+  cout<<"         "<<"--duration  duration in seconds"<< std::endl;
+}
+
+void string_to_map(const std::string& input, std::map<std::string, std::string>& saml_map)
+{
+  std::string token, key, value;
+  std::istringstream iss(input);
+
+  while (std::getline(iss, token, ',')) {
+  token.erase(std::remove_if( token.begin(), token.end(),
+    [](char c){ return (c =='\r' || c =='\t' || c == ' ' || c == '\n' || c == '{' || c == '}' || c == '"');}), token.end() );
+    auto pos = token.find(':');
+    key = token.substr(0, pos);
+    value = token.substr(pos + 1);
+    saml_map[key] = value;
+  }
+}
+
+int main(int argc,char *argv[]) {
+
+  vector<const char*> args;
+  argv_to_vec(argc, (const char **)argv, args);
+  env_to_vec(args);
+
+  auto cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+                         CODE_ENVIRONMENT_UTILITY, 0);
+
+  common_init_finish(g_ceph_context);
+
+  std::string role_id, policy, val, err;
+  int duration = 0;
+  std::map<std::string, std::string> saml_map;
+  for (std::vector<const char*>::iterator i = args.begin(); i != args.end(); ) {
+    if (ceph_argparse_flag(args, i, &val, "--help", (char*)NULL)) {
+      help();
+      return 0;
+    } else if (ceph_argparse_witharg(args, i, &val, "--role-id", (char*)NULL)) {
+      role_id = val;
+    } else if (ceph_argparse_witharg(args, i, &val, "--policy", (char*)NULL)) {
+      policy = val;
+    } else if (ceph_argparse_witharg(args, i, &val, "--duration", (char*)NULL)) {
+      duration = (int)strict_strtol(val.c_str(), 10, &err);
+    } else if (ceph_argparse_witharg(args, i, &val, "--saml-keys", (char*)NULL)) {
+      string_to_map(val, saml_map);
+    }
+  }
+  RGWSTS sts(duration, role_id, policy, saml_map);
+  sts.build_output();
+  struct timeval tv = sts.get_expiration_time();
+  std::string time_str = timeval_to_string(tv);
+  cout << "Credentials: Access Key Id: " << sts.get_access_key_id() << std::endl;
+  cout << "Credentials: Secret Access Key: " << sts.get_secret_access_key() << std::endl;
+  cout << "Credentials: Expiration: " << sts.get_expiration_time() << std::endl;
+  cout << "Credentials: Session Token: " << sts.get_session_token() << std::endl;
+  cout << "AssumedRoleUser: Assumed Role Id: " << sts.get_role_id() << std::endl;
+  cout << "PackedPolicySize: "  <<  sts.get_packed_policy_size() << std::endl;
+
+  return 0;
+}
+
```

Comment:
@adamemerson @mdw-at-linuxbox : Please re-review. I have used the formatter, as suggested by Marcus. Also, I am not using any compression for the STS token, because looks like the token will never be greater that size 65 using hmac-sha256. I have added error handling but the error codes are not exact - they can be replaced when we have the STS errors in place.

Also do we need to add more complexity to the access_key and secret_key generation, and maintain a list of (say last 10) credentials issued, so that the credentials are not repeated frequently. or should we depend on method to give us a random alphanumeric string?